### ISSUE TITLE
Guest customization script fails on ubuntu 18.4.3 LTS

### ISSUE THREAD

**Issue by deburau** [vmware/open-vm-tools#381]
In ubuntu 18.4.3 TLS the directory **/var/run** is a link to the directory **/run** which is mounted with the **noexec** flag:

`# mount|grep "/run "`
`tmpfs on /run type tmpfs (rw,nosuid,noexec,relatime,size=204168k,mode=755)`

But the guest customization script is executed in a subdirectory of **/var/run/** and fails with **Exit code: 32256**.

Because of this, the whole guest customization fails.

**Comment by xiaofengw-vmware** [comment]
Yes, the customization script need the execution privilege and if it is mounted as noexec, customization will fail. I have opened a PR to make this customization script dir configurable. Will update you when the  fix is available. Thanks.

**Comment by xiaofengw-vmware** [comment]
Hi deburau,
Could you share us the detail about the ubuntu 18.4.3? Is it a desktop/server/live server? We want to reproduce it. Thanks.

**Comment by deburau** [comment]
I used this ISO to install it: **ubuntu-18.04.3-live-server-amd64.iso** which was downloaded from _https://ubuntu.com/download/server/thank-you?country=GB&version=18.04.3&architecture=amd64_

**Comment by xiaofengw-vmware** [comment]
Thanks, I will download and have a try.

**Comment by deburau** [comment]
It is definitly the noexec flag. After I changed the system to mount **/run** without **noexec**, the customization script was executed.

This is what I changed:

In **/usr/share/initramfs-tools/init** I removed "noexec," from the line
`mount -t tmpfs -o "noexec,nosuid,size=10%,mode=0755" tmpfs /run`

Then I rebuilded the initramfs with

`update-initramfs -uv`

After that, the script was executed during forced guest customization.

**Comment by xiaofengw-vmware** [comment]
After that, the script was executed during forced guest customization.
[Xiaofeng]: As to the script, do you mean you input a custom script to run during the guest customization?  If yes, could you share me your vCenter version? And are you using cloud-init or traditional guest customization? Thanks.

**Comment by deburau** [comment]
vCloud Director version: 9.7.0.14534864 Fri Aug 30 13:29:55 CEST 2019
vSphere 6.7.0.21000

I'm using a custom customization script, the simplest case is just a "#".

Ubuntu 18.4.3 LTS uses cloud-init.

I've done a fresh install without any customizations (just using OpenSSH server). Then power off, setting script to "#" and power on and force recustomization. Result - guest customization failed.

I have attached the toolsDeployPkg.log
[toolsDeployPkg.log](https://github.com/vmware/open-vm-tools/files/3777972/toolsDeployPkg.log)

**Comment by xiaofengw-vmware** [comment]
Thanks deburau, since it's a vCD forced guest customization. vCD still consume guest customization 6.5/6.7 codes and I will talk with them about the solution. (crossport or consume new codes). I will update soon.

**Comment by xiaofengw-vmware** [comment]
vCD team decide to consume our vsphere 6.7 imgcust codes which fixed this issue and I will ask them to update this once this is done. Thanks.

**Comment by xiaofengw-vmware** [comment]
Hi Deburau,
Based on the process, please contact vCD GSS to open a SR, then GSS will open a PR which will be assigned to vCD team.  Thanks a lot.

**Comment by xiaofengw-vmware** [comment]
vSpehere65p04  patch branch is completed. I need to wait for the next patch branch to be opened.

**Comment by jidckii** [comment]
Hi @xiaofengw-vmware !
Any news on this bug?
It also doesn't work on ubuntu 20.04.

**Comment by alexandrovas** [comment]
Same error on Ubuntu 18.04.5 and open-vm-tools 11.0.5-4ubuntu0.18.04.1

**Comment by xiaofengw-vmware** [comment]
> Hi @xiaofengw-vmware !
> Any news on this bug?
> It also doesn't work on ubuntu 20.04.
Hi @jidckii , the vCD needs to be upgraded to a new version, I will ask vCD guys to provide a correct version which has already fixed this issue. Thanks.

**Comment by xiaofengw-vmware** [comment]
Hi @jidckii 
I just got response from vCD team, they want you to  open a SR against vCD GSS to get the build with the fix. Thanks.

**Comment by hjstam** [comment]
This issue is fixed in VMware Cloud Director version: 10.1.2.16779297. In this version of Cloud Director the customization script is executed in another directory (/root/.customization).

```
INFO: RunCustomScript invoked in '/var/run/.vmware-imgcust-dq9Yowq/' for 'precustomization'
DEBUG: Command: '/bin/cat /var/run/.vmware-imgcust-dq9Yowq//script3709331039159361666.bat | /usr/bin/tr -d '^M' > /var/run/.vmware-imgcust-dq9Yowq//script3709331039159361666.bat.tmp'
DEBUG: Result:
DEBUG: Exit Code: 0
DEBUG: Command: '/bin/mv -f /var/run/.vmware-imgcust-dq9Yowq//script3709331039159361666.bat.tmp /var/run/.vmware-imgcust-dq9Yowq//script3709331039159361666.bat'
DEBUG: Result:
DEBUG: Exit Code: 0
INFO: Making temporary post-customization directory
DEBUG: Command: '/bin/mkdir /root/.customization'
DEBUG: Result:
DEBUG: Exit Code: 0
DEBUG: Command: '/bin/cp /var/run/.vmware-imgcust-dq9Yowq//script3709331039159361666.bat /root/.customization/customize.sh'
DEBUG: Result:
DEBUG: Exit Code: 0
DEBUG: Command: '/bin/chmod u+x /root/.customization/customize.sh'
DEBUG: Result:
DEBUG: Exit Code: 0
INFO: Executing pre-customization script...
INFO: precustomization
DEBUG: Command: '/root/.customization/customize.sh "precustomization"'
DEBUG: Result:
DEBUG: Exit Code: 0
INFO: RunCustomScript has completed
```

I'v tested with Ubuntu 18.04.5 and open-vm-tools 11.0.5-4ubuntu0.18.04.1

**Comment by carmineesposito73** [comment]
Is this fix included in vCloud Director 9.7.0.6 for Service Providers | 12 NOV 2020 | Build 17135839 (installed build 17135848)?

**Comment by xiaofengw-vmware** [comment]
> 17135848

Yes, it is included in vCD ob-17135848.

**Comment by stanguturi** [comment]
As @xiaofengw-vmware mentioned in previous comments, the issue is fixed in the latest release. Closing this issue.
